newsrevenuehub/rev-engine
# DEV-1815 - Payment receipt emails are not rendering correctly

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

After PR is approved, final TODO is to update the [Sending NRE email Wiki](https://news-revenue-hub.atlassian.net/wiki/spaces/TECH/pages/2104786945/Sending+email+through+NRE) to reflect changes.

#### What's this PR do?

In order to fix the immediate problem of confirmation emails not properly rendering conditional logic in templates, this PR refactors our ~`send_email`~ `send_templated_email` task so that it renders the message body template on our end, instead of referencing a template stored on Mailgun, and sending required context to have Mailgun handle the rendering.

The need for this became apparent after I understood why the conditional blocks were not properly displaying in Mailgun. When I was working on [DEV-1056 - Deliver (initial) confirmation emails for recurring donations](https://news-revenue-hub.atlassian.net/browse/DEV-1056), I failed to appreciate that Mailgun uses the handlebars templating language.

While handlebars does support conditional blocks, in on order to test a condition (namely, is the contribution interval one time or not, which is what we do in our template), you need to be able register a custom Handlebars helper ([see, for instance, this answer on SO](https://stackoverflow.com/a/51976315)). However, Mailgun does not support registering custom Handlebar helpers.

Given that there is already a desire to remove templating logic from Mailgun altogether (in part, so we're not tightly coupled with this particular ESP), the best path forward is to fully render our email templates in the app, and only rely on the ESP for sending the message, not rendering templates.

In addition to fixing contribution confirmation emails, this PR also refactors magic email link donor emails as sent in `RequestContributorTokenEmailView.post` to render the template on our end, using `send_templated_email`.

These are the only two emails that get sent that rely on Anymail/Mailgun + templating.  Moving forward, if we require additional transactional emails, we will continue to rely on app-side (as opposed to ESP-side) template rendering.

#### Why are we doing this? How does it help us?

It fixes the acute problem of contribution confirmation emails not properly rendering and provides a more robust solution for transactional emails moving forward.

#### How should this be manually tested?

Make a contribution in environment and see that the email sent to you is properly rendered.

Request a magic email link, then open the email and click link. It should work.


#### Have automated unit tests been added?

Existing unit tests will be modified to account for changes.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

N/A

#### Has this been documented? If so, where?

This is a TODO in order for this PR to be considered complete. Need to update [Sending NRE email Wiki](https://news-revenue-hub.atlassian.net/wiki/spaces/TECH/pages/2104786945/Sending+email+through+NRE).

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-1815](https://news-revenue-hub.atlassian.net/browse/DEV-1815)
[DEV-1687](https://news-revenue-hub.atlassian.net/browse/DEV-1687)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them?

No next steps.
